### ISSUE TITLE
Immediate-mode simulation

### ISSUE THREAD

**Issue by Ralith** [dimforge/rapier#324]

**Comment by sebcrozet** [comment]
Transferred this issue to the `rapier` repository.

**Comment by PoseidonEnergy** [comment]
@Ralith OP, did you ever find a workaround for this in Rapier? I'm needing this functionality for exactly the reason in your post (game networking).

**Comment by Ralith** [comment]
No, this is still unsolved. You can approximate by manually computing where a body would end up without collisions, then doing a shape cast to that position and getting as close as you can, but the error this introduces is highly visible to users.

**Comment by PoseidonEnergy** [comment]
@Ralith Thanks for the response. Do you know of any other strategies for multiplayer networking that don't require "stepping ahead" individual objects?

**Comment by Ralith** [comment]
There's all sorts of tricks you can play, but it's all very application specific. Ultimately this functionality is needed for certain cases. If you're blocked on it, I'd encourage you to try implementing it.

**Comment by Vrixyz** [comment]
I'm curious to see real use-case of this feature, to help with understanding implications.

**Comment by Ralith** [comment]
I'm happy to discuss real-world use in detail. What part of the descriptions given above are unclear?

**Comment by upisfree** [comment]
> I'm curious to see real use-case of this feature, to help with understanding implications.

I'm developing a browser-based online game with Rapier as the physical engine and a server on node.js. The client (browser) and the server (node.js) run the same version of the physical simulation, for ease of state synchronization.

But, for example, for client prediction error correction I need to be able to update only one object rather than the whole world. For example, when a desynchronization occurs and I roll back the player to the last confirmed state from the server and apply all commands sent (since desynchronization) to move the player.

I am trying to implement [this system](https://developer.valvesoftware.com/wiki/Latency_Compensating_Methods_in_Client/Server_In-game_Protocol_Design_and_Optimization#Client_Side_Prediction).

**Comment by Ralith** [comment]
I'm glad you found success with an alternative approach, but per my initial description, the pattern of only predicting a few specific items is widespread, and the feature request stands.

**Comment by Ralith** [comment]
Pseudocode illustrating how an MVP implementation might be used, per discussion with @Vrixyz about this at RustWeek:

```rs
let mut time_remaining: f32;
let mut body: RigidBody;
while time_remaining > 0.0 {
    // Simulate the movement of `body` through the world until it collides with something
    let Some((time_of_impact, contacts)) = integrate_motion(&mut body, &colliders, time_remaining) else {
        break;
    };
    time_remaining -= time_of_impact;
    // Apply forces to `body`
    let forces = resolve_collision(&body, &contacts);
    body.apply_force(forces);
}
```
The key missing API here is `resolve_collision`. Applications can manage a pretty good approximation of `integrate_motion` by manually integrating forces/velocities and running shape casts across the resulting path at whatever granularity is desired, but a built-in option could probably be a bit better. Note that there is no `RigidBodySet` involved -- this is intentional. Only one body is moving, because the rest of the world state is already set in stone. The caller might optionally choose to apply reaction forces to contacted bodies in the next full-simulation time-step, or to apply collision filters such that the only possible interactions are with static bodies, or design for inconsistencies arising from missing reaction forces to just be difficult to observe.

A complete implementation would support immediate-mode simulation of multiple bodies at once, e.g. to support lag compensation for a player tossing some nun-chucks . PhysX's API (link in OP has been fixed) seems to do something like that. This would be great, but even an implementation that only supports a single body at a time would be a big step forwards.
